### PR TITLE
feat: lifi get final quote

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR removes the edge-case for Li.Fi, ensuring that like other swappers, gets a final trade quote with the most up-do-date data (rate/fees).
This improves things twofolds:

- First, for tokens with approval required, this ensures that users get the correct gas fee after approval has been granted (which we weren't updating before, since we were never getting a final trade quote)
- This also ensures that users are notified in case of rate updated

## Issue (if applicable)

- contributes to the closed https://github.com/shapeshift/web/issues/8271 - ensuring we not only get accurate gas estimates for native tokens/tokens with allowance at rate time, but also update gas at quote time for tokens with just-granted allowance
- contributes to https://github.com/shapeshift/web/issues/8270

## Risk

> High Risk PRs Require 2 approvals

Medium - if we get this wrong, Li.Fi final trade quote time would show an empty broken state, see testing steps below

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

⚠️ Before beginning testing, you have to try and get a route which you notice is unstable between quotes i.e:

- Try and get a quote for a given amount
- Add $0.01, then back down to reforce quotes refetch
- Notice a given Li.Fi quote will now be missing while the two others are present. That's your unstable quotes.

From my testing, 1 FOX.ARB -> ETH.ARB seems to be a good contender, as ZRX/Li.Fi and Enso Li.Fi routes will be present in some quotes list, but not consistently:

Screenshot 2024-12-10 at 17 06 37
Screenshot 2024-12-10 at 17 07 56

- Once you found your unstable Li.Fi quote, go all the way to final trade quote and ensure you don't end up in this state:

Screenshot 2024-12-10 at 17 08 09

Note, you'll have to try to back out to input, then back to final quotes many many times (while having the quote you have spotted as unstable selected) over to ensure we consistently get said route and don't end up with the empty quote bug above

### Engineering

- Ensure we pass the allowed tool in Li.Fi final trade quote req as `exchanges: { allow: [tool] }`

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

- ☝🏽 

## Screenshots (if applicable)

- Li.Fi swapper consistently gets final trade quote without early return null with no quote bug

https://jam.dev/c/fc162032-bbff-4c89-927c-259c79686071

- Li.Fi network fees are accurate on final quote after allowance has been granted

https://jam.dev/c/1e1ea531-5961-49b3-b53f-4b8c54eefe09

### CODE REVIEW

File: packages/swapper/src/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts [hunk 1]
```diff
@@ -243,10 +247,11 @@ export async function getTrade(
 
       return {
         id: selectedLifiRoute.id,
-        // This isn't a mistake - with Li.Fi, we can never go with our full-on intent of rate vs. quotes. As soon as a wallet is connected, we get a *quote*
-        // even though we're lying and saying this is a rate. With the "rate" containing a receiveAddress, a quote will *not* be fired at pre-sign time, which
-        // ensures users aren't rugged with routes that aren't available anymore when going from input to confirm
-        receiveAddress,
+        // TODO(gomes): when https://github.com/shapeshift/web/pull/8309 goes in, this goes out
+        // We do need receiveAddress in *input* to send it as fromAddress for routes req for more reliable rates, but with receiveAddress currently being the quotes/rates discriminator,
+        // we need to exclude it from method in *output*
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts:252]: i.e whichever goes in first will have to be rebase and mergefix to accomodate for the new `quoteOrRate` discriminator, while keeping the `lifiTools` diff

File: src/state/apis/swapper/swapperApi.ts [hunk 2]
```diff
@@ -91,15 +91,6 @@ export const swapperApi = createApi({
...
                 ...tradeQuoteInput,
-                // Receive address should always be undefined for trade *rates*, however, we *do* pass it to check for cross-account support
-                // so we have to ensure it is gone by the time we call getTradeRates
-                receiveAddress:
-                  // Ok, we may have lied just before. Li.Fi is the odd one, and we don't want to fetch a final quote for it.
-                  // The reason is Li.Fi routes are very fragile and from one request to the other, it's more likely than not a tool may not be returned anymore, resulting in early null returns
-                  // in swapper, i.e a totally broken state.
-                  // By keeping the receiveAddress in all the way through, and returning it in LIFI's `getTradeRate()`, the rate becomes an effective quote, and the
-                  // `shouldFetchTradeQuote = isExecutableTradeQuote(activeTrade)` never becomes true, meaning we keep the rate as a quote.
-                  swapperName === SwapperName.LIFI ? tradeQuoteInput.receiveAddress : undefined,
```
Comment (woodenfurniture) [inline src/state/apis/swapper/swapperApi.ts:102]: :crazyeyes:

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx [hunk 3]
```diff
@@ -289,9 +275,8 @@ export const useGetTradeQuotes = () => {
         sellAccountType: sellAccountMetadata?.accountType,
         buyAsset,
         wallet: wallet ?? undefined,
-        // This isn't a mistake, we're not fetching anything.
-        // Li.Fi rate is an actual quote and we want to leverage cache, see swapperApi's `getTradeQuote` for more details
-        quoteOrRate: swapperName === SwapperName.LIFI ? 'rate' : 'quote',
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:294]: ❤️

**Review by woodenfurniture — COMMENTED** [review]
# Testing notes

✅ Unable to get into state described - working even with unstable quote

⚠️ Unable to actually do the trade due to quite severe rate limit ing from lifi. Its unclear whether this is related to the changes here, as if caching or something else is affected then possibly it is?



https://jam.dev/c/b529a925-920c-44c7-b50e-f6bbd4bf453f

**Review by woodenfurniture — APPROVED** [review]
Ok, retested after not spamming the UI and trades do in fact go through fine

https://jam.dev/c/ad447ac2-8891-4969-9565-dba0f609b1db

**Pull request merged**